streaming/receiver.c

```diff
@@ -51,47 +162,264 @@ static void rrdpush_receiver_thread_cleanup(void *ptr) {
+    PARSER_USER_OBJECT *user = user_v;
     UNUSED(plugins_action);
-    char *remote_time_txt = words[1];
-    time_t remote_time = 0;
-    RRDHOST *host = ((PARSER_USER_OBJECT *)user)->host;
-    struct plugind *cd = ((PARSER_USER_OBJECT *)user)->cd;
-    if (cd->version < VERSION_GAP_FILLING ) {
-        error("STREAM %s from %s: Child negotiated version %u but sent TIMESTAMP!", host->hostname, cd->cmd,
-               cd->version);
-        return PARSER_RC_OK;    // Ignore error and continue stream
-    }
-    if (remote_time_txt && *remote_time_txt) {
-        remote_time = str2ull(remote_time_txt);
-        time_t now = now_realtime_sec(), prev = rrdhost_last_entry_t(host);
-        time_t gap = 0;
-        if (prev == 0)
-            info("STREAM %s from %s: Initial connection (no gap to check), remote=%ld local=%ld slew=%ld",
-                 host->hostname, cd->cmd, remote_time, now, now-remote_time);
-        else {
-            gap = now - prev;
-            info("STREAM %s from %s: Checking for gaps... remote=%ld local=%ld..%ld slew=%ld  %ld-sec gap",
-                 host->hostname, cd->cmd, remote_time, prev, now, remote_time - now, gap);
+    char *id = words[1];
+    char *start_txt = words[2];
+    char *first_txt = words[3];
+    char *end_txt = words[4];
+    if (!id || !start_txt || !end_txt)
+        goto disable;
+
+    RRDSET *st = rrdset_find(user->host, id);
+    if (unlikely(!st))
+        goto disable;
+    user->st = st;
+    st->state->window_start = str2ll(start_txt, NULL);
+    st->state->window_first = str2ll(first_txt, NULL);
+    st->state->window_end   = str2ll(end_txt, NULL);
+
+    struct receiver_state *rpt = user->opaque;
+    time_t now = now_realtime_sec();
+    if (st->last_updated.tv_sec == 0) {
+        if (rpt->gap_history==0) {
+            st->last_updated.tv_sec = st->state->window_start;
+            debug(D_REPLICATION, "Initial data for %s, no historical gaps. window=%ld..%ld",
+                  st->name, (long)st->state->window_start, (long)st->state->window_end);
+            st->state->ignore_block = 0;
+            return PARSER_RC_OK;
+        } else {
+            debug(D_REPLICATION, "Initial data for %s, asking for history window=%ld..%ld",
+                  st->name, (long)now - rpt->gap_history+1, now);
+            enqueue_replication_req(rpt, user->host, st->id, now - rpt->gap_history + 1, now);
+            st->state->ignore_block = 1;
+            st->last_updated.tv_sec = now - rpt->gap_history;
+            return PARSER_RC_OK;
         }
-        char message[128];
-        sprintf(message,"REPLICATE %ld %ld\n", remote_time - gap, remote_time);
-        int ret;
-#ifdef ENABLE_HTTPS
-        SSL *conn = host->stream_ssl.conn ;
-        if(conn && !host->stream_ssl.flags) {
-            ret = SSL_write(conn, message, strlen(message));
+    }
+    time_t expected_t = st->last_updated.tv_sec + st->update_every;
```

Comment:
I think it would be better do not use suffix `_t` to name variables, this is expected for data type as you can see in its own declaration ( `time_t` ).